diff --git a/lib/png_decoder.js b/lib/png_decoder.js
index 58578698eb498022dbfc9d8dcbeaaaeb21f78995..4ca6a503b3e63a79941fa02d0a5f6169c0b1a434 100644
--- a/lib/png_decoder.js
+++ b/lib/png_decoder.js
@@ -284,8 +284,9 @@ export default class PngDecoder extends IOBuffer {
                 if (length % 2 !== 0) {
                     throw new RangeError(`tRNS chunk length must be a multiple of 2. Got ${length}`);
                 }
-                if (length / 2 > this._png.width * this._png.height) {
-                    throw new Error(`tRNS chunk contains more alpha values than there are pixels (${length / 2} vs ${this._png.width * this._png.height})`);
+                const expectedLength = this._colorType === ColorType.GREYSCALE ? 2 : 6;
+                if (length !== expectedLength) {
+                    throw new Error(`tRNS chunk must contain ${expectedLength} bytes for this color type. Got ${length}`);
                 }
                 this._hasTransparency = true;
                 this._transparency = new Uint16Array(length / 2);
diff --git a/src/png_decoder.ts b/src/png_decoder.ts
index 7f27123caac37442fcd0348f77a651cfaab3b73b..5078deada3e95ab1829a2c70df93a13bc11e236d 100644
--- a/src/png_decoder.ts
+++ b/src/png_decoder.ts
@@ -329,11 +329,10 @@ export default class PngDecoder extends IOBuffer {
             `tRNS chunk length must be a multiple of 2. Got ${length}`,
           );
         }
-        if (length / 2 > this._png.width * this._png.height) {
+        const expectedLength = this._colorType === ColorType.GREYSCALE ? 2 : 6;
+        if (length !== expectedLength) {
           throw new Error(
-            `tRNS chunk contains more alpha values than there are pixels (${
-              length / 2
-            } vs ${this._png.width * this._png.height})`,
+            `tRNS chunk must contain ${expectedLength} bytes for this color type. Got ${length}`,
           );
         }
         this._hasTransparency = true;
